pvs/560 A part of conditional expression is always false [skip lint]

ngortheone opened the pull request:
2 commits here:
1. Lint in regex.c linted regmatch function

* comment style
* if/esle braces
* some post-increment instead of pre-increment

Lint still fails, reporting very long function, and complains about pre-decrement `     --rp->rs_un.regsave.rs_u.pos.lnum`, hence `skip-lint`

2.  removing always true condition `status != RA_FAIL` according to pvs report. Reason:
 before changed line status variable is set to `RA_FAIL` in 2 places line `3816` and `3826`, both lead to short-circuit of the loop, making comparing it to `RA_FAIL` useless

ngortheone committed:
lint: regmatch

* comment style
* if/esle braces
* post-increment instead of pre-increment

ngortheone commented on src/nvim/regexp.c:
```diff
@@ -4829,33 +4860,35 @@ static int regmatch(
         if (status == RA_NOMATCH) {
           reg_restore(&rp->rs_un.regsave, &backpos);
-          --brace_count[rp->rs_no];             /* decrement match count */
+          brace_count[rp->rs_no]--;             // decrement match count
```
increment changed here. not sure if semantics are preserved

ngortheone commented on src/nvim/regexp.c:
```diff
@@ -4829,33 +4860,35 @@ static int regmatch(
           reg_restore(&rp->rs_un.regsave, &backpos);
-          --brace_count[rp->rs_no];
-          /* continue with the items after "\{}" */
+          brace_count[rp->rs_no]--;
```
increment changed here. not sure if semantics are preserved

justinmk replied:
they are. But this file is listed in [frozen modules](https://github.com/neovim/neovim/wiki/Refactoring) because we don't plan to change it significantly. Thus we don't want to spend time on it. In particular, linting it is a low priority.

ngortheone commented on src/nvim/regexp.c:
```diff
@@ -4941,16 +4974,16 @@ static int regmatch(
                   || reg_getline(
-                      --rp->rs_un.regsave.rs_u.pos.lnum)
-                  == NULL)
+                      (rp--)->rs_un.regsave.rs_u.pos.lnum)
```
increment changed here. not sure if semantics are preserved

justinmk replied:
this is wrong. Did the linter actually complain about this line?

ngortheone replied:
I can omit all increment related changes, leaving only braces and spaces

ngortheone replied:
I reverted this one

ngortheone commented:
The pvs fix is not here yet. I decided to go over lints first

On July 27, 2019 2:44:04 PM EDT, "Justin M. Keyes" <notifications@github.com> wrote:
>justinmk commented on this pull request.
>
>
>
>>                if (rp->rs_un.regsave.rs_u.pos.lnum
>                   < behind_pos.rs_u.pos.lnum
>                   || reg_getline(
>-                      --rp->rs_un.regsave.rs_u.pos.lnum)
>-                  == NULL)
>+                      (rp--)->rs_un.regsave.rs_u.pos.lnum)
>
>this is wrong. Did the linter actually complain about this line?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/10624#discussion_r307974037

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

blueyed commented:
> The pvs fix is not here yet. I decided to go over lints first

As for CI you could use `[skip lint]` - if this / the linter gets in your way.

ngortheone commented:
The was trying to fix this https://neovim.io/doc/reports/pvs/PVS-studio.html.d/sources/regexp.c_8.html#ln4326 but I still struggle to understand why pvs things that.

ngortheone committed:
pvs/V560: part of conditional expression is always true

* before changed line status variable is set to RA_FAIL in 2 places
line 3816 and 3826, both lead to short-circuit of the loop, making
comparing it to RA_FAIL useless

blueyed commented:
@ngortheone 
> ngortheone changed the title pvs/560 A part of conditional expression is always false pvs/560 A part of conditional expression is always false [skip lint]

JFI: Needs to be in the commit message.. but OTOH it is also not necessary to run the second stage then anyway.
I often end up editing .travis.yml/appveyor.yml to include only one job or similar.. https://github.com/neovim/neovim/pull/10610 is meant to make that easier (i.e. only run the single coverage jobs)

justinmk commented on src/nvim/regexp.c:
```diff
@@ -4353,7 +4353,7 @@ static int regmatch(
           }
 
-          if (status != RA_FAIL && status != RA_NOMATCH) {
+          if (status != RA_NOMATCH) {
```
if we "know" it's always true then it should be asserted. then we avoid future regressions.

    assert(status != RA_FAIL);

justinmk commented:
Merged without the lint changes.
